lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -480,18 +490,35 @@ class MongoClientInterface {
     putObject(bucketName, objName, objVal, params, log, cb) {
         MongoUtils.serialize(objVal);
         const c = this.getCollection(bucketName);
+        const countParams = {
+            objName,
+            objVal,
+        };
+        const mongoCall = {
+            args: [c, bucketName, objName, objVal, params, log],
+            cb,
+        };
+        let isUpdate = false;
         if (params && params.versioning && !params.versionId) {
-            return this.putObjectVerCase1(c, bucketName, objName, objVal,
-                                          params, log, cb);
+            countParams.isVer = true;
+            mongoCall.op = this.putObjectVerCase1.bind(this);
         } else if (params && params.versionId === '') {
-            return this.putObjectVerCase2(c, bucketName, objName, objVal,
-                                          params, log, cb);
+            countParams.isVer = false;
+            mongoCall.op = this.putObjectVerCase2.bind(this);
         } else if (params && params.versionId) {
-            return this.putObjectVerCase3(c, bucketName, objName, objVal,
-                                          params, log, cb);
+            countParams.versionId = params.versionId;
+            mongoCall.op = this.putObjectVerCase3.bind(this);
+            isUpdate = true;
+        } else {
+            countParams.isVer = false;
+            mongoCall.op = this.putObjectNoVer.bind(this);
+        }
+        if (isUpdate) {
+            return dataCounterHelper.updateObjWrapper(
+            this.dataCount, c, countParams, log, mongoCall);
         }
-        return this.putObjectNoVer(c, bucketName, objName, objVal,
-                                   params, log, cb);
+        return dataCounterHelper.addObjWrapper(
```

Comment:
This inverts control, even if it's a bit more repetitive the following pattern:

- do operation
- if no error, then account for it

would be more readable and would avoid conditioning the mongo call to the wrapper not failing. Also it incurs a lot of arguments serialization that can be a performance penalty.